rust/src/storage/s3/dynamodb.rs

```diff
@@ -0,0 +1,427 @@
+            vars::OWNER_NAME_PATH.to_string() => OWNER_NAME.to_string(),
+            vars::IS_RELEASED_PATH.to_string() => IS_RELEASED.to_string(),
+        };
+        let mut values = hashmap! {
+            vars::IS_RELEASED_VALUE.to_string() => attr("1"),
+            vars::RVN_VALUE.to_string() => attr(&lock.record_version_number),
+            vars::OWNER_NAME_VALUE.to_string() => attr(&lock.owner_name),
+        };
+
+        let update: &str;
+        if let Some(ref data) = lock.data {
+            update = expressions::UPDATE_IS_RELEASED_AND_DATA;
+            names.insert(vars::DATA_PATH.to_string(), DATA.to_string());
+            values.insert(vars::DATA_VALUE.to_string(), attr(data));
+        } else {
+            update = expressions::UPDATE_IS_RELEASED;
+        }
+
+        self.client
+            .update_item(UpdateItemInput {
+                table_name: self.opts.table_name.clone(),
+                key: hashmap! {
+                    PARTITION_KEY_NAME.to_string() => attr(self.opts.partition_key_value.clone())
+                },
+                condition_expression: Some(
+                    expressions::PK_EXISTS_AND_OWNER_RVN_MATCHES.to_string(),
+                ),
+                update_expression: Some(update.to_string()),
+                expression_attribute_names: Some(names),
+                expression_attribute_values: Some(values),
+                ..Default::default()
+            })
+            .await?;
+        Ok(())
+    }
+
+    async fn upsert_item(
+        &self,
+        data: Option<String>,
+        condition_expression: Option<String>,
+        expression_attribute_names: Option<HashMap<String, String>>,
+        expression_attribute_values: Option<HashMap<String, AttributeValue>>,
+    ) -> Result<LockItem, DynamoError> {
+        let lookup_time = now_millis();
+        let rvn = Uuid::new_v4().to_string();
+
+        let mut item = hashmap! {
+            PARTITION_KEY_NAME.to_string() => attr(self.opts.partition_key_value.clone()),
+            OWNER_NAME.to_string() => attr(&self.opts.owner_name),
+            RECORD_VERSION_NUMBER.to_string() => attr(&rvn),
+            LEASE_DURATION.to_string() => attr(&self.opts.lease_duration),
+        };
+
+        if let Some(ref d) = data {
+            item.insert(DATA.to_string(), attr(d));
+        }
+
+        self.client
+            .put_item(PutItemInput {
+                table_name: self.opts.table_name.clone(),
+                item,
+                condition_expression,
+                expression_attribute_names,
+                expression_attribute_values,
+                ..Default::default()
+            })
+            .await?;
+
+        Ok(LockItem {
+            owner_name: self.opts.owner_name.clone(),
+            record_version_number: rvn,
+            lease_duration: self.opts.lease_duration,
+            is_released: false,
+            data,
+            lookup_time,
```

Comment:
that's how it is in the origin version, I believe they address the clock skew here

Reply:
yeap, that's the case https://github.com/awslabs/amazon-dynamodb-lock-client#how-we-handle-clock-skew

Reply:
Sorry for the confusion, let me try to clarify my original question with more context. The lookup time is recorded before we set the relative lease duration time in dynamodb. The way I understand this is lease duration starts counting after the item in dymoadb is created. However, by recording the lookup time before we send the dynamodb request, we are kicking off the expiration timer earlier than we should, which could cause false positive expiration detection.

I noticed this when I was looking at our `get_lock` implementation, where we record lookup time after the dynamodb request. This might result in longer longer time to detect expiration (false negative), but at least we can guarantee correctness. I took a quick look at `updateItemAndStartSessionMonitor` in the java implementation, they also record the lookup time after the dynamodb request.

Reply:
> The lookup time is recorded before we set the relative lease duration time in dynamodb 

This is where item is put in dynamo https://github.com/delta-io/delta-rs/pull/153/files#diff-cdce0c557ea383d6919df0813d5e8f29bf38aa2efa60dcda197b94386f102b57R322-R327, e.g. `lookup_time` is local value and not stored in dynamodb. Moreover, none of the possible timestamps are on dynamodb.

Regarding your next assumption 
> This might result in longer time to detect expiration
yes, that is true, although with that we guarantee that the lock will not be acquire _before_ the expiration under any condition

Reply:
We might revisit that by ensuring that there won't be clock skew and work with timestamp in dynamodb but with less safe environment

Reply:
However we talk here only for non expected scenarios as when the lock is not released, which means that worker which held it had died. We might tune lease_duration later once we establish possible fault scenarios

Reply:
> > This might result in longer time to detect expiration
>
> yes, that is true, although with that we guarantee that the lock will not be acquire before the expiration under any condition

Yep, what i meant in my previous reply was I think this behavior is OK since it guarantees correctness. However, this doesn't seem to be the behavior for `upsert_item` because local timer might expire sooner than the configured lease duration. Basically I am proposing that we do:

```rust
        // dyamodb put_item request
        self.client.put_item(...);

        Ok(LockItem {
            owner_name: self.opts.owner_name.clone(),
            record_version_number: rvn,
            lease_duration: self.opts.lease_duration,
            is_released: false,
            data,
            now_millis(),
        })
```

Instead of what we have right now:

```rust
        let lookup_time = now_millis();
 
        // dyamodb put_item request
        self.client.put_item(...);

        Ok(LockItem {
            owner_name: self.opts.owner_name.clone(),
            record_version_number: rvn,
            lease_duration: self.opts.lease_duration,
            is_released: false,
            data,
            lookup_time,
        })
```

This should make the local timer behavior consistent with our `get_lock` and upstream's `updateItemAndStartSessionMonitor` implementation.

Reply:
oh wow, that's true